beluga/include/beluga/motion/ackerman_drive_model.hpp

```diff
@@ -0,0 +1,269 @@
+    const auto delta_time = std::chrono::duration<double>(time - previous_time).count();
+    if constexpr (std::is_same_v<state_type, Sophus::SE2d>) {
+      return sampling_fn_2d(pose, previous_pose, delta_time);
+    } else {
+      return sampling_fn_3d(pose, previous_pose, delta_time);
+    }
+  }
+
+ private:
+  using control_type_2d = std::tuple<Sophus::SE2d, Sophus::SE2d>;
+  using control_type_3d = std::tuple<Sophus::SE3d, Sophus::SE3d>;
+
+  [[nodiscard]] auto sampling_fn_3d(const Sophus::SE3d& pose, const Sophus::SE3d& previous_pose, double delta_time)
+      const {
+    const auto current_pose_2d = To2d(pose);
+    const auto previous_pose_pose_2d = To2d(previous_pose);
+    const auto two_d_sampling_fn = sampling_fn_2d(current_pose_2d, previous_pose_pose_2d, delta_time);
+    return [=](const state_type& state, auto& gen) { return To3d(two_d_sampling_fn(To2d(state), gen)); };
+  }
+
+  [[nodiscard]] auto sampling_fn_2d(const Sophus::SE2d& pose, const Sophus::SE2d& previous_pose, double delta_time)
+      const {
+    // Calculate velocities from poses
+    const auto [linear_velocity, angular_velocity] = calculate_velocities(pose, previous_pose, delta_time);
+
+    using DistributionParam = typename std::normal_distribution<double>::param_type;
+
+    // Velocity noise parameters (following velocity motion model from Probabilistic Robotics)
+    const auto linear_velocity_params = DistributionParam{
+        linear_velocity, std::sqrt(
+                             params_.translation_noise_from_translation * std::abs(linear_velocity) +
+                             params_.translation_noise_from_rotation * std::abs(angular_velocity))};
+
+    const auto angular_velocity_params = DistributionParam{
+        angular_velocity, std::sqrt(
+                              params_.rotation_noise_from_translation * std::abs(linear_velocity) +
+                              params_.rotation_noise_from_rotation * std::abs(angular_velocity))};
+
+    // Additional orientation noise (gamma_hat) using rotation parameters
+    const auto gamma_params = DistributionParam{
+        0.0,  // zero mean
+        std::sqrt(
+            params_.orientation_noise_from_translation * std::abs(linear_velocity) +
+            params_.orientation_noise_from_rotation * std::abs(angular_velocity))};
+
+    return [=](const auto& state, auto& gen) {
+      static thread_local auto distribution = std::normal_distribution<double>{};
+
+      // Sample noisy velocities
+      const auto v_hat = distribution(gen, linear_velocity_params);
+      const auto omega_hat = distribution(gen, angular_velocity_params);
+      const auto gamma_hat = distribution(gen, gamma_params);
+
+      // Apply velocity motion model
+      return apply_velocity_motion(state, v_hat, omega_hat, gamma_hat, delta_time);
+    };
+  }
+
+  /// Calculate linear and angular velocities from two poses and delta time
+  std::pair<double, double>
```

Comment:
@fbattocchia nit: consider using a custom `struct` for this. `v` and `w` read better than `first` and `second`.